Add CPFP

jnewbery opened the pull request:
Added a section on Child Pays For Parent

jnewbery commented:
@jamesob @moneyball - any review greatly appreciated!

harding commented on 1.fee_bumping/fee_bumping.md:
```diff
@@ -177,21 +179,128 @@ TODO
+transaction B in the block, she must also include transaction A, since without
+A, B is spending a non-existent output and is invalid.
+
+If a miner considered transactions independently when constructing her block,
```
Nit: s/a miner/the miner

jnewbery replied:
fixed

harding commented on 1.fee_bumping/fee_bumping.md:
```diff
@@ -177,21 +179,128 @@ TODO
+transaction input, divided by the transaction size). Bitcoin Core [PR #1590][]
+changed the mining code to sort transactions by fee rate. [Version 0.7.0][],
+released in September 2012 was therefore the first Bitcoin Core release with
+any fee-optimized behavior for miners.
```
AFAIK, the preceeding text about ordering transactions by fee is correct, however I think you're using that information to draw the conclusion that 0.7 was the first release with fee-optimized behavior---which I think is an incorrect conclusion.

Nakamoto introduced code in 0.3.something, I believe, that filled the block template in priority order (as described) but it also increased the minimum fee for a transaction to be added to the block template as the template approached half of what was called then the "soft maximum block size" (which was 500KB, so 250KB was half).  The increases were exponential, so e.g. the first 100KB could be cheap or free but the last 50KB could be quite pricey.

This is not fee optimal---for that you need feerate sorting with CPFP as we have now[1]---but it is fee optimized.  If you need, I can find that Nakamoto code.

[1] Although to be truly optimal, you probably need to have infinitely spacious mempools and consider CPFP tx chains of infinite depth.  Happily, good approximations of optimal can be achieved much more simply.

jnewbery replied:
You're good at this! Yes, Satoshi added code here to raise the fee required as the block filled up: https://github.com/bitcoin/bitcoin/commit/f1e1fb4bdef878c8fc1564fa418d44e7541a7e83#diff-23cfe05393c8433e384d2c385f06ab93R584, and then updated it here: https://github.com/bitcoin/bitcoin/commit/9b8eb4d6907502e9b1e74b62a850a11655d50ab5#diff-23cfe05393c8433e384d2c385f06ab93R587.

You're right about the current strategy being optimized, but not optimal. Maximizing the fee yield in a block is a tricky problem. Even if there we knew a better-than-brute-force method to make transaction selection optimal, it'd need to be efficient. There's not much use in an algorithm that finds the optimal block if it takes ten minutes to do so!

harding commented on 1.fee_bumping/fee_bumping.md:
```diff
@@ -177,21 +179,128 @@ TODO
+
+The Bitcoin Core mining code was updated [in 2016][CPFP PR] to better account
+for packages of transactions. The mining code will consider packages of up to
+25 transactions or 101vkB. This changes was included in [V0.13][].
```
Nit: s/changes/change/

jnewbery replied:
fixed

jnewbery commented:
Thanks for the review @harding (and especially the history lesson on block construction :slightly_smiling_face: ). I think I've addressed all your points.

I agree that diagrams would be helpful. I'll look at adding those in a future PR.

bitschmidty commented on 1.fee_bumping/fee_bumping.md:
```diff
@@ -67,7 +67,7 @@ enough fee, then miners will be incentivized to include it in a block.
 ##### Child Pays For Parent (CPFP) 
 
-The user creates  a new transaction which spends one or more of the outputs of
+The user creates a new transaction which spends one or more of the outputs of
```
createsa spelling

jnewbery replied:
Fixed. Thanks!

jnewbery commented on 1.fee_bumping/fee_bumping.md:
```diff
@@ -177,21 +179,128 @@ TODO
+## How does CPFP work?
+
+When constructing a new block, miners are incentivized to fill the 1vMB with
+the set of transactions that maximize their miner fee. If all unconfirmed
```
I've settled on 'transaction fees'. Thanks!

jnewbery commented on 1.fee_bumping/fee_bumping.md:
```diff
@@ -177,21 +179,132 @@ TODO
+
+## How does CPFP work?
+
+When constructing a new block, miners are incentivized to fill the 1vMB with
```
Added as a footnote for now (and since github flavored markdown doesn't support footnotes, it's just a note saying "look at footnote). I think as the scaling book gets expanded, we're going to use vB all over the place, so I've also left a comment to move the footnote to a more appropriate place in the book later.

jnewbery commented on 1.fee_bumping/fee_bumping.md:
```diff
@@ -177,21 +179,132 @@ TODO
+which took into account the transaction fee of children transaction when
+sorting transactions for inclusion in a block. This patch was used by at least
+some miners, but was never merged into Bitcoin Core due to a lack of testing
+and benchmarking, and concerns that it could open a DOS vector against miners.
```
fixed

jnewbery commented on 1.fee_bumping/fee_bumping.md:
```diff
@@ -177,21 +179,132 @@ TODO
+confirmed and gets stuck in the mempool. This leads to customers complaining
+about slow or stuck transactions.
+
+To improve the experience for their customrs, HBE implemented CPFP to bump the
```
fixed
